src/Swaps/Swapper.sol

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+pragma solidity ^0.8.0;
+
+import {IEVault, IERC20} from "evk/EVault/IEVault.sol";
+import {SafeERC20Lib} from "evk/EVault/shared/lib/SafeERC20Lib.sol";
+import {RevertBytes} from "evk/EVault/shared/lib/RevertBytes.sol";
+
+import {ISwapper} from "./ISwapper.sol";
+
+import {OneInchHandler} from "./handlers/OneInchHandler.sol";
+import {UniswapV2Handler} from "./handlers/UniswapV2Handler.sol";
+import {UniswapV3Handler} from "./handlers/UniswapV3Handler.sol";
+import {UniswapAutoRouterHandler} from "./handlers/UniswapAutoRouterHandler.sol";
+
+/// @title Swapper
+/// @custom:security-contact security@euler.xyz
+/// @author Euler Labs (https://www.eulerlabs.com/)
+/// @notice Untrusted helper contract for EVK for performing swaps and swaps to repay
+contract Swapper is OneInchHandler, UniswapV2Handler, UniswapV3Handler, UniswapAutoRouterHandler {
+    bytes32 public constant HANDLER_ONE_INCH = bytes32("1Inch");
+    bytes32 public constant HANDLER_UNISWAP_V2 = bytes32("UniswapV2");
+    bytes32 public constant HANDLER_UNISWAP_V3 = bytes32("UniswapV3");
+    bytes32 public constant HANDLER_UNISWAP_AUTOROUTER = bytes32("UniswapAutoRouter");
+
+    uint256 internal constant REENTRANCYLOCK_UNLOCKED = 1;
+    uint256 internal constant REENTRANCYLOCK_LOCKED = 2;
+
+    uint256 private reentrancyLock;
+
+    error Swapper_UnknownMode();
+    error Swapper_UnknownHandler();
+    error Swapper_Reentrancy();
+    error Swapper_InsufficientBalance();
+    error Swapper_PastDeadline();
+
+    // In the locked state, allow contract to call itself, but block all external calls
+    modifier externalLock() {
+        bool isExternal = msg.sender != address(this);
+
+        if (isExternal) {
+            if (reentrancyLock == REENTRANCYLOCK_LOCKED) revert Swapper_Reentrancy();
+            reentrancyLock = REENTRANCYLOCK_LOCKED;
+        }
+
+        _;
+
+        if (isExternal) reentrancyLock = REENTRANCYLOCK_UNLOCKED;
+    }
+
+    constructor(address oneInchAggregator, address uniswapRouterV2, address uniswapRouterV3, address uniswapRouter02)
+        OneInchHandler(oneInchAggregator)
+        UniswapV2Handler(uniswapRouterV2)
+        UniswapV3Handler(uniswapRouterV3)
+        UniswapAutoRouterHandler(uniswapRouter02)
+    {}
+
+    /// @inheritdoc ISwapper
+    function swap(SwapParams memory params)
+        public
+        externalLock
+    {
+        if (params.mode >= MODE_MAX_VALUE) revert Swapper_UnknownMode();
+        if (params.deadline < block.timestamp) revert Swapper_PastDeadline();
+
```

Comment:
maybe call `setMaxAllowance` here rather in each dedicated handler?

Reply:
only handlers know which contract to grant allowance to. In theory there could be push based handlers.